### DISCUSSION TITLE
Release build library that uses TCA results in a runtime crash inside debug application

### DISCUSSION THREAD

**Discussion by mrtnlst** [pointfreeco/swift-composable-architecture#2363] [General]
### Description

I'm facing an obscure crash when building a static library that uses TCA (only links, not embeds) in a _release_ configuration to customers. When the framework is added to an app target (alongside adding TCA) and building it in a _debug_ configuration, a crash occurs in [WithViewStore's initializer](https://github.com/pointfreeco/swift-composable-architecture/blob/fb561a37fc94564d1b80b634875b8bad51e6f33b/Sources/ComposableArchitecture/SwiftUI/WithViewStore.swift#L122). If we remove all code inside the `#if DEBUG` compiler directive there, it works.

If this is an issue of TCA or something entirely else we don't know. 
We have faced this issue since at least version 0.35.1.

![Crash](https://github.com/mrtnlst/SampleProject/blob/5d0ed571495d96f708b311d53bd2e22855e59a96/Resources/Screenshot%20of%20Crash.png?raw=true)

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

I expect that a _release_ built binary framework (that uses TCA as a dependency) can be build and run inside an app in either _debug_ or _release_ configuration. 

### Actual behavior

A _release_ built binary framework (that uses TCA as a dependency) can only be run inside an app in _release_ configuration.

### Steps to reproduce

I have provided an example project at https://github.com/mrtnlst/SampleProject.
It consists of: 
* a Framework target using a SwiftUI `View` and `WithViewStore` that has TCA as a _Target Dependency_
* an App target that embeds the framework and also embeds TCA

By default the project embeds the framework using the + button under _Frameworks, Libraries, and Embedded Content_ and you can see it builds and launches successfully (displaying a button).

To reproduce the crash: 
1. Remove the embedded framework under _Frameworks, Libraries, and Embedded Content_ 
2. Select the Framework scheme and build it (it's already configured to run in _release_ configuration)
3. Select the product in the project navigator below "Products"
4. Open the `SampleProjectFramework` in Finder
5. Drag the `SampleProjectFramework` inside the framework artefact to  _Frameworks, Libraries, and Embedded Content_
6. Build and run the App target again

### Hints
The issue doesn't occur when 
* The framework is linked and embedded into the app target **via Xcode** (the build configuration propagates from app to framework I guess)
* The framework target is build with the _debug_ configuration or the app target runs with the _release_ configuration (no mismatch between the 2)
* We comment out all Code inside [WithViewStore's initializer](https://github.com/pointfreeco/swift-composable-architecture/blob/fb561a37fc94564d1b80b634875b8bad51e6f33b/Sources/ComposableArchitecture/SwiftUI/WithViewStore.swift#L122) that is inside the `#if DEBUG` compiler directive

![Removed content inside compiler directive](https://github.com/mrtnlst/SampleProject/blob/main/Resources/Removed%20DEBUG%20condition%20content.png?raw=true)

### The Composable Architecture version information

1.0.0

### Destination operating system

iOS 16

### Xcode version information

Xcode 14.3

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)
```

**Comment by mbrandonw** [comment]
Hi @mrtnlst, have you confirmed that this is actually a problem with TCA, or just a problem in general Swift projects when importing release libraries into debug projects? I'm not aware of anything we can do specifically in the library to prevent this kind of crash. I would imagine this happens in any Swift project.

And beyond that, we highly recommend against doing this in general. Building TCA for release causes all types of warnings, preconditions and helpers to be omitted from the library. You will no longer be notified when the library detects improper usage of an API, the `_printChanges` operator will not work, and worst of all `TestStore` will not work at all so you can't even write tests.

Since we're not really sure this is an actual TCA problem I am going to convert this to a discussion as we all try to figure out what is going on.

**Reply by mrtnlst** [reply]
Hi @mbrandonw, can you elaborate what you mean with "we highly recommend against doing this"? 

For my case, I don't compile TCA for release and provide it to users, just my own framework. Of course my framework uses TCA as a dependency and therefor I have to link it to TCA while compiling. TCA is then declared as a dependency in the Package.swift of my framework, that I provide to users. And I expect them to build both for debug as well, to take benefit of the mentioned benefit by you. 

Or do you consider this not as a normal use case? 

*EDIT*:
Okay I just recreated the whole setup without TCA, but instead created a local package, that is linked to my static framework and both are embedded into the app. It all works, but as soon as I add `#if DEBUG` conditions into the local package's code, a similar looking crash appears. I have never seen something like this 🤯 

So that means a distributed release binary framework and an embedding app can't share a dependency (on the same version) if there are `DEBUG` conditions in it?
